### PR TITLE
1 RxJs

### PR DESCRIPTION
**Pull request by Yuri-9**
image

1. Home task: [Home-task-RxJs](https://d17btkcdsmqrmh.cloudfront.net/docs/module_1-intro-to-rxjs/home-task)
2.  Total (5/5)
- Add Input handler **0.5**
- Filter Input value  **0.5**
- Add debounce time **1**
- Combining the results of requests **1**
- Add loader logic to our page **1**
- Unsubscribe from all subscriptions **1**

### CODE REVIEW

File: src/app/app.component.ts [hunk 1]
```diff
@@ -47,15 +49,30 @@ export class AppComponent implements OnInit, OnDestroy {
     this.charactersResults$ = this.searchTermByCharacters
       .pipe
       // YOUR CODE STARTS HERE
-
-      // YOUR CODE ENDS HERE
-      ();
+      (tap((v) => console.log(v)),
```
Comment (MishkanJigan) [inline src/app/app.component.ts]: `console.log(v)` - I think that's unnecessary.

File: src/app/app.component.ts [hunk 2]
```diff
@@ -47,15 +49,30 @@ export class AppComponent implements OnInit, OnDestroy {
...
+      })
+    );
+    // YOUR CODE ENDS HERE
   }
 
   loadCharactersAndPlanet(): void {
     // 4. On clicking the button 'Load Characters And Planets', it is necessary to process two requests and combine the results of both requests into one result array. As a result, a list with the names of the characters and the names of the planets is displayed on the screen.
     // Your code should looks like this: this.planetAndCharactersResults$ = /* Your code */
     // YOUR CODE STARTS HERE
+    this.planetAndCharactersResults$ = this.mockDataService
```
Comment (MishkanJigan) [inline src/app/app.component.ts]: Try to use `combineLatest `operator here.

File: src/app/app.component.ts [hunk 3]
```diff
@@ -66,12 +83,28 @@ export class AppComponent implements OnInit, OnDestroy {
     - Subscribe to changes
     - Check the received value using the areAllValuesTrue function and pass them to the isLoading variable. */
     // YOUR CODE STARTS HERE
+    const arrayLoaders$ = this.mockDataService.getPlanetLoader().pipe(
```
Comment (MishkanJigan) [inline src/app/app.component.ts]: Try to use combineLatest operator here.

File: src/app/app.component.ts [hunk 4]
```diff
@@ -66,12 +83,28 @@ export class AppComponent implements OnInit, OnDestroy {
...
+    });
+
     // YOUR CODE ENDS HERE
   }
 
   ngOnDestroy(): void {
     // 5.2 Unsubscribe from all subscriptions
     // YOUR CODE STARTS HERE
+
+    this.searchTermByCharacters.unsubscribe();
```
Comment (MishkanJigan) [inline src/app/app.component.ts]: You have more subscriptions that you forgot about.
Also, try to use `pipe(takeUntil())` constraction when you subscribe on observable.

File: src/app/app.component.ts [hunk 5]
```diff
@@ -66,12 +83,28 @@ export class AppComponent implements OnInit, OnDestroy {
     - Subscribe to changes
     - Check the received value using the areAllValuesTrue function and pass them to the isLoading variable. */
     // YOUR CODE STARTS HERE
+    const arrayLoaders$ = this.mockDataService.getPlanetLoader().pipe(
+      switchMap((isPlanetLoading) =>
+        this.mockDataService.getCharactersLoader().pipe(
+          map((isPCharacterLoading) => {
+            return [!isPlanetLoading, !isPCharacterLoading];
```
Comment (MishkanJigan) [inline src/app/app.component.ts]: why do you use here (!)?

Comment (MishkanJigan) [reply src/app/app.component.ts]: I understand what this is for, if you want, you'd better correct the `areAllValuesTrue()` function.

File: src/app/app.component.ts [hunk 6]
```diff
@@ -66,12 +83,28 @@ export class AppComponent implements OnInit, OnDestroy {
...
+          })
+        )
+      )
+    );
+    arrayLoaders$.subscribe((loaders) => {
```
Comment (MishkanJigan) [inline src/app/app.component.ts]: When you will use `combineLatest()` you can immediately subscribe to it and thus get rid of the `arrayLoaders `variable, but don't forget about the `takeUntil()` **unsubscribe**.

File: src/app/app.component.ts (hunk 1 above)
Comment (Yuri-9) [reply src/app/app.component.ts]: fixed

File: src/app/app.component.ts (hunk 2 above)
Comment (Yuri-9) [reply src/app/app.component.ts]: fixed

File: src/app/app.component.ts (hunk 3 above)
Comment (Yuri-9) [reply src/app/app.component.ts]: fixed

File: src/app/app.component.ts (hunk 4 above)
Comment (Yuri-9) [reply src/app/app.component.ts]: fixed

File: src/app/app.component.ts (hunk 5 above)
Comment (Yuri-9) [reply src/app/app.component.ts]: fixed

File: src/app/app.component.ts (hunk 6 above)
Comment (Yuri-9) [reply src/app/app.component.ts]: fixed

File: src/app/app.component.ts [hunk 7]
```diff
@@ -47,15 +51,25 @@ export class AppComponent implements OnInit, OnDestroy {
...
 
   loadCharactersAndPlanet(): void {
     // 4. On clicking the button 'Load Characters And Planets', it is necessary to process two requests and combine the results of both requests into one result array. As a result, a list with the names of the characters and the names of the planets is displayed on the screen.
     // Your code should looks like this: this.planetAndCharactersResults$ = /* Your code */
     // YOUR CODE STARTS HERE
+
+    this.planetAndCharactersResults$ = combineLatest([
+      this.mockDataService.getCharacters(),
+      this.mockDataService.getPlanets(),
+    ]).pipe(map(([characters, planets]) => characters.concat(planets)), takeUntil(this.destroy$));
```
Comment (MishkanJigan) [inline src/app/app.component.ts]: This is unnecessary here - takeUntil(this.destroy$)

File: src/app/app.component.ts [hunk 8]
```diff
@@ -66,16 +80,30 @@ export class AppComponent implements OnInit, OnDestroy {
     - Subscribe to changes
     - Check the received value using the areAllValuesTrue function and pass them to the isLoading variable. */
     // YOUR CODE STARTS HERE
+
+    combineLatest([
+      this.mockDataService.getPlanetLoader(),
+      this.mockDataService.getCharactersLoader(),
+    ]).
+    pipe(takeUntil(this.destroy$)).
+    subscribe(
```
Comment (MishkanJigan) [inline src/app/app.component.ts:89]: One small thing, you also can do like this, if you want
```
subscribe(
      (loaderStates: boolean[]) =>
        (this.isLoading = this.isSomeValuesTrue(loaderStates))
    )
```

Comment (Yuri-9) [reply src/app/app.component.ts:89]: it is a good idea

File: src/app/app.component.ts (hunk 7 above)
Comment (Yuri-9) [reply src/app/app.component.ts]: remoded

**Pull request merged**